### ISSUE TITLE
Restricting fields access

### ISSUE THREAD

**Issue by dessHub** [fratzinger/feathers-casl#8]
Hi!

I'm trying to switch from `feathers-permissions` to CASL and I came across `feathers-casl`. One of the features I wanted to leverage from CASL is restricting some fields from being updated. 

# Expected behaviour:
An `agency admin` should only be able to update `roles` field and nothing else.

# My rules definition:
![Screenshot 2021-02-08 at 12 26 35](https://user-images.githubusercontent.com/19705477/107200568-ed647b00-6a08-11eb-8acb-7818730bf4c9.png)

NB: `[angency-users]` is the service.

# Current behaviour

It allows update for all the fields but restrict the results to only `roles`. 

My question is, how can I restrict update to only `roles` and deny the others? What I'm I doing wrong?

**Comment by fratzinger** [comment]
Hey :) Thanks for the good explanation!

There are tests for this scenario: https://github.com/fratzinger/feathers-casl/blob/0e7c4e04ff9c835c9466dce3f5733a35201f5846/test/authorize/authorize.complete.test.ts#L621

It's hard to say from your small snippet. Could it come from other rules?

**Comment by dessHub** [comment]
Yes this if from the rules. 
![Screenshot 2021-02-08 at 14 28 52](https://user-images.githubusercontent.com/19705477/107213818-0f1a2e00-6a1a-11eb-9af9-d6a98fd5efc3.png)

**Comment by fratzinger** [comment]
This will not work:
```
{ agencyId: agencyIds });
{ id: agencyIds }
```

I think it should be (which is not your problem):
````
{ agencyId: { $in: agencyIds } });
{ id: { $in: agencyIds } }
```

Could you set a breakpoint and look which rules the corresponding ability have?

**Comment by dessHub** [comment]
This didn't fix it
`{ agencyId: { $in: agencyIds } });
{ id: { $in: agencyIds } }`

This is what's in the `ability`
![Screenshot 2021-02-08 at 14 57 31](https://user-images.githubusercontent.com/19705477/107216711-307d1900-6a1e-11eb-94cb-26d9c42d773c.png)

**Comment by fratzinger** [comment]
It doesn't help much, as the conditions are not displayed properly.
But the rule you talk about (`action: 'update'`, `subject: 'agency-users'` is not there at all).

Walk through your `defineRulesFor` function and make sure your logic is correctly.

**Comment by dessHub** [comment]
Actually the rule is there
![Screenshot 2021-02-08 at 14 57 31](https://user-images.githubusercontent.com/19705477/107218333-881c8400-6a20-11eb-829a-65cb474bcc2d.png)

**Comment by fratzinger** [comment]
sorry, my bad! But it's hard to say where the error lies.

**Comment by dessHub** [comment]
Okay thank you, let me still play around with it. For some reasons its seems to be restricting the response instead of the inputs.

**Comment by dessHub** [comment]
@fratzinger I think I might know where the issue is. On `patch`, `context.id` is `null`. And in your library, there is this [condition](https://github.com/fratzinger/feathers-casl/blob/0e7c4e04ff9c835c9466dce3f5733a35201f5846/lib/hooks/authorize/authorize.hook.before.ts#L86). Therefore, it doesn't get [here](https://github.com/fratzinger/feathers-casl/blob/0e7c4e04ff9c835c9466dce3f5733a35201f5846/lib/hooks/authorize/authorize.hook.before.ts#L117)

**Comment by fratzinger** [comment]
You're totally right! There's something wrong. Your code should get [here](https://github.com/fratzinger/feathers-casl/blob/0e7c4e04ff9c835c9466dce3f5733a35201f5846/lib/hooks/authorize/authorize.hook.before.ts#L142).

I'm sorry, I currently have no quick solution for this. Just for now you could get around this by not using `multi`-patch but that's maybe not possible for you.
I'll work on this and get back to you, when I got something.

Sorry for the inconvenience! `feathers-casl` is at a early stage.

**Comment by dessHub** [comment]
No worries, I will try find a work around. Thank you

**Comment by fratzinger** [comment]
This should be fixed in version `^0.1.8-0`.
Please try it again with `npm i feathers-casl@0.1.8-0` and give feedback. It's a pre-release. but it's more stable than the current version. Hopefully I will bring this to `v1` soon.

Please also notice that this is major upgrade from casl@4 to casl@5. You might come across errors, which should be fixable by following this: https://github.com/stalniy/casl/releases/tag/%40casl%2Fangular%405.0.0

**Comment by dessHub** [comment]
Thanks @fratzinger , I will check and give an update.

**Comment by dessHub** [comment]
@fratzinger it works. It doesn't throw a forbidden error though.

**Comment by fratzinger** [comment]
That is as expected. I think it could be an option if that should pass or not.

**Comment by dessHub** [comment]
yeah that make sense.

**Comment by fratzinger** [comment]
Closing this in favor of #11. Please feel free to comment, if there's something else :)

**Comment by fratzinger** [comment]
released as stable with `^0.2.0`. See release notes: https://github.com/fratzinger/feathers-casl/releases/tag/v0.2.0
